5.drug-dependency/scripts/4.drug_latent_dim_correlation.py

```diff
@@ -0,0 +1,263 @@
+    latent_df.set_index('ModelID', inplace=True)
+
+    # Align both dataframes based on the ModelID
+    common_model_ids = latent_df.index.intersection(drug_df.index)
+    # Filter both dataframes to keep only common ModelIDs
+    latent_df_filtered = latent_df.loc[common_model_ids]
+    prism_df_filtered = drug_df.loc[common_model_ids]
+    # Check the variance of each latent dimension and drug response column
+    latent_variance = latent_df_filtered.var()
+    prism_variance = prism_df_filtered.var()
+
+    # Filter out constant columns (variance == 0)
+    latent_df_filtered = latent_df_filtered.loc[:, latent_variance != 0]
+    prism_df_filtered = prism_df_filtered.loc[:, prism_variance != 0]
+
+    # Loop over each latent dimension and calculate correlation with each drug
+    for latent_col in latent_df_filtered.columns:
+        for drug_col in prism_df_filtered.columns:
+            latent_values = latent_df_filtered[latent_col]
+            drug_values = prism_df_filtered[drug_col]
+
+            # Check if either column is constant
+            if latent_values.nunique() <= 1 or drug_values.nunique() <= 1:
+                corr = np.nan
+            else:
+                # Drop missing values for both columns
+                valid_data = pd.concat([latent_values, drug_values], axis=1).dropna()
+                latent_values_valid = valid_data[latent_col]
+                drug_values_valid = valid_data[drug_col]
+
+                if len(latent_values_valid) > 1 and len(drug_values_valid) > 1:
+                    # Calculate Pearson correlation
+                    corr, p_value = pearsonr(latent_values_valid, drug_values_valid)
+                else:
+                    corr = np.nan
+                    print("nan")
+        
+                # Store the results
+                result_row = {
+                    "z": int(latent_col.replace("z_", "")),
+                    "full_model_z": num_components,
+                    "model": str(model_name),
+                    "drug": str(drug_col),
+                    "pearson_correlation": corr,
+                    "p_value": p_value,
+                    "shuffled": False
+                }
+                correlation_results.append(result_row)
+    
+    # Convert results into a dataframe
+    correlation_results_df = pd.DataFrame(correlation_results)
+    return correlation_results_df
+
+
+# In[3]:
+
+
+data_directory = pathlib.Path("../0.data-download/data").resolve()
+dependency_file = pathlib.Path(f"{data_directory}/CRISPRGeneEffect.parquet").resolve()
+gene_dict_file = pathlib.Path(f"{data_directory}/CRISPR_gene_dictionary.parquet").resolve()
+
+
+# In[4]:
+
+
+# Load PRISM data
+top_dir = "../5.drug-dependency"
+data_dir = "data"
+
+prism_df, prism_cell_df, prism_trt_df = load_utils.load_prism(
+    top_dir=top_dir,
+    data_dir=data_dir,
+    secondary_screen=False,
+    load_cell_info=True,
+    load_treatment_info=True,
+)
```

Comment:
I love the keyname args here! makes the function call inuitive

Reply:
Thanks!